nomad/client_rpc.go

```diff
@@ -65,7 +69,20 @@ func (s *Server) removeNodeConn(ctx *RPCContext) {
 
 	s.nodeConnsLock.Lock()
 	defer s.nodeConnsLock.Unlock()
-	delete(s.nodeConns, ctx.NodeID)
+	state, ok := s.nodeConns[ctx.NodeID]
+	if !ok {
+		return
+	}
+
+	// It is important that we check that the connection being removed is the
+	// actual stored connection for the client. It is possible for the client to
+	// dial various addresses that all route to the same server. The most common
+	// case for this is the original address the client uses to connect to the
+	// server differs from the advertised address sent by the heartbeat.
+	if state.Ctx.Conn.LocalAddr().String() == ctx.Conn.LocalAddr().String() &&
+		state.Ctx.Conn.RemoteAddr().String() == ctx.Conn.RemoteAddr().String() {
+		delete(s.nodeConns, ctx.NodeID)
```

Comment:
Is it required that this node id is always present in the stored connections? What if it is not?